lib/ansible/modules/cloud/kubevirt/kubevirt_vm.py

```diff
@@ -249,92 +256,99 @@ def argspec(self):
+        else:
+            want_created = want_ready = False
+
+        for event in self._vm_resource.watch(namespace=self.namespace, timeout=self.params.get('wait_timeout')):
+            entity = event['object']
+            if entity.metadata.name != self.name:
+                continue
+            status = entity.get('status', {})
+            created = status.get('created', False)
+            ready = status.get('ready', False)
+            if (created, ready) == (want_created, want_ready):
+                return entity
+
+        self.fail("Error waiting for virtual machine. Try a higher wait_timeout value.")
+
+    def manage_state(self, new_state, already_changed):
+        new_running = True if new_state == 'running' else False
+        changed = False
+        k8s_obj = {}
+
+        if not already_changed:
+            k8s_obj = self.get_resource(self._vm_resource)
+            if not k8s_obj:
+                self.fail("VirtualMachine object disappeared during module operation, aborting.")
+            if k8s_obj.spec.get('running', False) == new_running:
+                return False, k8s_obj
+
+            newdef = dict(metadata=dict(name=self.name, namespace=self.namespace), spec=dict(running=new_running))
+            k8s_obj, err = self.patch_resource(self._vm_resource, newdef, k8s_obj,
+                                               self.name, self.namespace, merge_type='merge')
+            if err:
+                self.fail_json(**err)
             else:
-                self._manage_state(False, resource_vm, existing, wait, wait_timeout)
-                return True
+                changed = True
+
+        if self.params.get('wait'):
+            k8s_obj = self._wait_for_state(new_state)
+
+        return changed, k8s_obj
 
     def execute_module(self):
         # Parse parameters specific for this module:
-        definition = virtdict()
         ephemeral = self.params.get('ephemeral')
         state = self.params.get('state')
-
-        if not ephemeral:
-            definition['spec']['running'] = state == 'running'
-
-        # Execute the CURD of VM:
-        template = definition['spec']['template']
-        kind = 'VirtualMachineInstance' if ephemeral else 'VirtualMachine'
-        dummy, definition = self.construct_vm_definition(kind, definition, template)
-        result = self.execute_crud(kind, definition)
-        changed = result['changed']
-
-        # Manage state of the VM:
-        if state in ['running', 'stopped']:
-            if not self.check_mode:
-                ret = self.manage_state(state)
-                changed = changed or ret
+        if ephemeral and state not in ('present', 'absent'):
+            raise TypeError("Ephemeral VMs only support 'present' and 'absent' states.")
+
+        _used_params = [name for name in self.params if self.params[name] is not None]
+        vm_spec_change = True if set(VM_SPEC_PARAMS).intersection(_used_params) else False
```

Comment:
What about inline stuff?

Reply:
You mean `resource_definition`? That's contained within VM_SPEC_PARAMS.

Reply:
Yeah, but does this code works well, when updating some attribute via that? As far as I read it correctly it doesn't.

Reply:
It does. If you call the module with any parameter that *might* modify spec: (and that includes `resource_definition`, which doesn't have to modify spec:, but it *might*), that's enough to trigger a full crud call. Only if the module is certain that none of the arguments used touch spec: in any way, does it skip the crud call.

I just pushed a new version rebased on top of master, which includes the module params cleanup PR and that has a much simpler `VM_SPEC_PARAMS` definition, so maybe that'll be clearer.

Reply:
I see now. Quite complex.. If we want to say user he can't stop/start non-existing VM, wouldn't it be simpler to just say that and instead of saying that he sent empty parameters tell him, which parameters are required(returned from backend) this would very simplify all of this and I don't think we really need this. But if you insist this is better I am ok.

Reply:
It's not that complex, the whole logic is ~10 lines of additional code. And the benefit is that in the 99% of cases where you just intend to start/stop a vm (like [1] where there are multiple other  examples)  you get a human–friendly error that clearly hints you're probably have your name wrong. Were it 100 lines of code, I wouldn't be so sure, but in this case the UX benefit is clear and the additional code isn't much.

[1] https://github.com/kubevirt/ansible-kubevirt-modules/blob/master/tests/playbooks/e2e.yaml#L41

Reply:
I am fine. It's just that if with too many and\or's are just too much for me :) I just preffer straighforward code with no complex ifs.